### ISSUE TITLE
`refresh=true` items should still run `shell` and `href` actions first

### ISSUE THREAD

**Issue by matryer** [matryer/xbar#626]
from https://github.com/matryer/xbar/issues/622

**Comment by matryer** [comment]
@seanw and @sydsyd - please see if this [new release](https://github.com/matryer/xbar/releases/latest) fixes your problems.

**Comment by sydsyd** [comment]
No, not for me. Is there a debug log I can send you?

**Comment by matryer** [comment]
@sydsyd - please try this release: [xbar.v2.0.22-beta-3-g528a063b.tar.gz](https://github.com/matryer/xbar/files/6123003/xbar.v2.0.22-beta-3-g528a063b.tar.gz)

If you run xbar like this: https://gist.github.com/matryer/51c4a186649a332cae2986d6ff44bb50 it will print out useful information which you can copy and paste.

**Comment by sydsyd** [comment]
% cd /Applications/xbar.app/Contents/MacOS
./xbar
2021/03/11 23:43:16 dnscrypt-proxy-switcher.10s.sh: /Users/admin/Library/Application Support/xbar/plugins/dnscrypt-proxy-switcher.10s.sh.vars.json, no such file:
There was a Javascript error. Please open the devtools for more information.
2021/03/11 23:43:21 dnscrypt-proxy-switcher.10s.sh: cycling: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:43:26 dnscrypt-proxy-switcher.10s.sh: refreshing: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:43:26 dnscrypt-proxy-switcher.10s.sh: cycling: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:43:32 dnscrypt-proxy-switcher.10s.sh: cycling: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:43:37 dnscrypt-proxy-switcher.10s.sh: refreshing: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:43:37 dnscrypt-proxy-switcher.10s.sh: cycling: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:43:42 dnscrypt-proxy-switcher.10s.sh: cycling: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:43:44 dnscrypt-proxy-switcher.10s.sh: exec: ./dnscrypt-proxy-switcher.10s.sh 9.9.9.9
2021/03/11 23:43:45 dnscrypt-proxy-switcher.10s.sh: ERR: action shell: fork/exec ./dnscrypt-proxy-switcher.10s.sh: no such file or directory
2021/03/11 23:43:45 dnscrypt-proxy-switcher.10s.sh: action refresh
2021/03/11 23:43:45 dnscrypt-proxy-switcher.10s.sh: refreshing: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:43:45 dnscrypt-proxy-switcher.10s.sh: cycling: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:43:50 dnscrypt-proxy-switcher.10s.sh: cycling: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:43:55 dnscrypt-proxy-switcher.10s.sh: cycling: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:43:55 dnscrypt-proxy-switcher.10s.sh: refreshing: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:43:57 dnscrypt-proxy-switcher.10s.sh: exec: ./dnscrypt-proxy-switcher.10s.sh empty
2021/03/11 23:43:57 dnscrypt-proxy-switcher.10s.sh: ERR: action shell: fork/exec ./dnscrypt-proxy-switcher.10s.sh: no such file or directory
2021/03/11 23:43:57 dnscrypt-proxy-switcher.10s.sh: action refresh
2021/03/11 23:43:57 dnscrypt-proxy-switcher.10s.sh: refreshing: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:43:57 dnscrypt-proxy-switcher.10s.sh: cycling: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:44:02 dnscrypt-proxy-switcher.10s.sh: cycling: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:44:07 dnscrypt-proxy-switcher.10s.sh: cycling: dnscrypt-proxy-switcher.10s.sh
2021/03/11 23:44:07 dnscrypt-proxy-switcher.10s.sh: refreshing: dnscrypt-proxy-switcher.10s.sh

**Comment by matryer** [comment]
Hmm, so it looks like `dnscrypt-proxy-switcher.10s.sh` can't be found. This could be because xbar allows you to install multiple plugins of the same name, so prefixes them with `001-` numbers.

I wonder it you could edit the plugin for me, replacing `dnscrypt-proxy-switcher.10s.sh` with `$0`, which should allow it to refer to itself properly, regardless of its name.

**Comment by sydsyd** [comment]
Not sure what you mean by edit the plugin - do you mean somewhere in the source?

**Comment by matryer** [comment]
Yeah, if you open the plugin (xbar menu, Preferences > Open Plugin) to open it in the app, you can click the **EDIT** button.

Then you can modify the plugin itself.

**Comment by sydsyd** [comment]
But I'm not sure where you want me to sub in $0
Here's the first few lines of the plugin

#!/usr/bin/env ksh
#
# dnscrypt-proxy switcher
#
# <bitbar.author.github>jedisct1</bitbar.author.github>
# <bitbar.author>Frank Denis</bitbar.author>
# <bitbar.desc>Toggle dnscrypt-proxy usage</bitbar.desc>
# <bitbar.image>https://raw.githubusercontent.com/jedisct1/bitbar-dnscrypt-proxy-switcher/master/bitbar-dnscrypt-proxy.jpg</bitbar.image>
# <bitbar.title>dnscrypt-proxy switcher</bitbar.title>
# <bitbar.url>https://github.com/jedisct1/bitbar-dnscrypt-proxy-switcher</bitbar.url>
# <bitbar.version>v0.1</bitbar.version>

# Theme: classic or emoji
THEME="classic"

# Non-authenticated resolver IP addresses
ADDITIONAL_IPS="9.9.9.9"

# Name of the non-authenticated resolver
ADDITIONAL_NAME="quad9"

# IP address of the proxy
DNSCRYPT_PROXY_IPS="127.0.0.1"

# -- do not change anything below this line --

if test -t 0; then
	THEME="emoji"
fi

case "$THEME" in
classic)

**Comment by matryer** [comment]
Actually, it already does it: 
![image](https://user-images.githubusercontent.com/101659/110791625-53dbf380-826a-11eb-8307-1ce3b9a18be2.png)

The "${0}" should work. Strange, not sure why you're getting that error then.

**Comment by matryer** [comment]
@sydsyd I think I've solved it. The problem was, the shelled command was being run with an unexpected working directory. I've fixed it so it is set properly, along with the environment variables, and wonder if you could test for me?

[xbar.v2.0.22-beta-8-gfb1c436d.tar.gz](https://github.com/matryer/xbar/files/6123187/xbar.v2.0.22-beta-8-gfb1c436d.tar.gz)

**Comment by sydsyd** [comment]
I think you're very close

**Comment by matryer** [comment]
What behaviour are you seeing now?

**Comment by sydsyd** [comment]
Nominal!

**Comment by matryer** [comment]
@sydsyd So, we're winning?

**Comment by sydsyd** [comment]
Looks that way!

**Comment by matryer** [comment]
Great. Closing despite @sydsyd's underwhelming responses. :)

**Comment by ggrelet** [comment]
> If you run xbar like this: https://gist.github.com/matryer/51c4a186649a332cae2986d6ff44bb50 it will print out useful information which you can copy and paste.

I'm humbly suggesting to add that tip on the README somewhere. This was really helpful for debugging a problem I had!

**Comment by goodyonsen** [comment]
> @sydsyd I think I've solved it. The problem was, the shelled command was being run with an unexpected working directory. I've fixed it so it is set properly, along with the environment variables, and wonder if you could test for me?
> 
> [xbar.v2.0.22-beta-8-gfb1c436d.tar.gz](https://github.com/matryer/xbar/files/6123187/xbar.v2.0.22-beta-8-gfb1c436d.tar.gz)

Hi, I installed xbar through the .dmg file.
I ran the xbar from the terminal properly. With only two apps installed the recurring output is; 

`2021/10/01 04:07:33 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:07:34 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh
2021/10/01 04:07:38 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:07:39 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh
2021/10/01 04:07:43 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:07:44 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh
2021/10/01 04:07:48 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:07:49 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh
2021/10/01 04:07:53 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:07:54 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh
2021/10/01 04:07:58 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:07:59 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh
2021/10/01 04:08:03 arp-spoof-detector.sh: refreshing: 001-arp-spoof-detector.sh
2021/10/01 04:08:03 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:08:04 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh
2021/10/01 04:08:08 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:08:09 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh
2021/10/01 04:08:13 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:08:14 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh
2021/10/01 04:08:18 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:08:19 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh
2021/10/01 04:08:23 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:08:24 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh
2021/10/01 04:08:28 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:08:29 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh
2021/10/01 04:08:33 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:08:34 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh
2021/10/01 04:08:38 arp-spoof-detector.sh: cycling: 001-arp-spoof-detector.sh
2021/10/01 04:08:39 dnsswitcher.1d.sh: cycling: 001-dnsswitcher.1d.sh`
 
But when I switch from the bar menu I get the following errors:
 
`2021/10/01 04:08:42 dnsswitcher.1d.sh: ERR: action shell: fork/exec /tmp/bitbar_dns_switcher_google: exec format error
`
`2021/10/01 04:08:45 dnsswitcher.1d.sh: ERR: action shell: fork/exec /tmp/bitbar_dns_switcher_level3: exec format error
`

Should I uninstall the app and trash the dmg and install the tar one you recommend?
